### PR TITLE
Fix redis host resolution

### LINKED ISSUE #12: Redis doesn't work with hostname
## Problem
The current release relies on [apicast:v3.0.0-alpha2](https://github.com/3scale/apicast-xc/blob/master/Dockerfile-apicast#L3). However, in this release the nginx `resolver` directive was [removed](https://github.com/3scale/apicast/blob/master/CHANGELOG.md#removed-2), and APIcast built-in resolver has to be used for all connections.

apicast-xc module doesn't take this into account, so as a result, Redis connection will only be created successfully when IP address is used in `XC_REDIS_HOST`. If hostname is used, it will fail. The client will only see a `403` error, and the reason returned by the Redis library is `no resolver defined to resolve "<redis hostname>"`.

## What is needed
Redis pool used by apicast-xc should use APIcast built-in resolver for Redis hostname.

### PR DESCRIPTION
**Pull request by mayorova**
Close #12

### CODE REVIEW

**Comment by mayorova** [comment]
As a further improvement I'd suggest reusing APIcast's function from [threescale_utils](https://github.com/3scale/apicast/blob/v3.0.0-beta3/apicast/src/threescale_utils.lua#L123), as it works with full Redis URL, with authorization and DB. But it's better to extract it to a separate module first.

**Comment by davidor** [comment]
I don't think that deleting the part of the code that overrides `header_filter` and `body_filter` is a good idea.
```
_M.header_filter = function() end		
_M.body_filter = function() end
```

That part of the code documents explicitly that XC does not want to perform any work in any of those 2 phases. If we delete that, every time that Apicast releases a new version, we will need to check whether it added new things in any of those two methods.

File: xc/cache.lua [hunk 1]
```diff
@@ -35,9 +35,10 @@ end
 -- @return true if authorized, false if denied, nil if unknown
 -- @return reason why the authorization is denied (optional)
 function _M.authorize(service_id, credentials, usage_method)
-  local redis, ok = redis_pool.acquire()
+  local redis, ok, err = redis_pool.acquire()
 
   if not ok then
+    ngx.log(ngx.WARN, "[cache] couldn't connect to redis: ", err)
```
Comment (davidor) [inline xc/cache.lua]: We should log a similar error in `_M.report`.

Comment (davidor) [reply xc/cache.lua]: If we do, we should change this error message to a more specific one, so the user knows whether the error happened while authorizing or while reporting.

**Comment by mayorova** [comment]
@davidor OK, agreed re: `header_filter` and `body_filter`, I'll just remove this commit from the PR

File: xc/cache.lua (hunk 1 above)
Comment (davidor) [reply xc/cache.lua]: Also, we should add a test that checks that we log a warning when there's an error connecting to Redis.
That test should be under this describe clause: `describe('when there is an error acquiring a Redis connection', function()` in `cache_spec.lua`.

File: xc/priority_auths.lua [hunk 2]
```diff
@@ -8,40 +8,49 @@ local _M = { }
 -- @return true if authorized, false if denied, nil if unknown
 -- @return reason why the authorization is denied (optional, required only when denied)
 function _M.authorize(service_id, credentials, metric)
-  local redis_pub, ok_pub = redis_pool.acquire()
+  local redis_pub, ok_pub, err = redis_pool.acquire()
 
   if not ok_pub then
+    ngx.log(ngx.WARN, "[priority auths] couldn't connect to redis pub: ", err)
```
Comment (davidor) [inline xc/priority_auths.lua:14]: I was going to say that we need tests for this too. However, I realized that `priority_auths` has no tests. That's homework for me and @unleashed :D

File: .busted [hunk 3]
```diff
@@ -0,0 +1,9 @@
+return {
```
Comment (davidor) [inline .busted]: We are currently sending the options in the `script/test` file, but I like the idea of using this file for clarity.
Apart from that, is the `lpath` option needed?

File: .busted [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+  _all = {
+    helper = 'spec/spec_helper.lua'
+  },
+  default = {
+    verbose = true,
+    lpath = "./spec/?.lua;"
+  },
+}
```
Comment (davidor) [inline .busted]: Missing newline.

File: spec/spec_helper.lua [hunk 5]
```diff
@@ -0,0 +1,7 @@
+local busted = require('busted')
+
+busted.before_each(function()
+  ngx = {
+    log = function() end
+  }
+end)
```
Comment (davidor) [inline spec/spec_helper.lua]: Missing newline.

File: .busted (hunk 3 above)
Comment (mayorova) [reply .busted]: I just stole this from APIcast :) https://github.com/3scale/apicast/blob/master/.busted
No idea how it works, to be honest.

File: xc/priority_auths.lua (hunk 2 above)
Comment (davidor) [inline xc/priority_auths.lua:14]: All these calls look the same: `ngx.log(ngx.WARN, error_prefix, err)`. How about extracting a `log_warning(error_prefix, err)` local method or something similar?

Also, I think that extracting all these error prefixes to a local constant would improve readability. Something like: `local ERRORS = { redis_pub_connect: "[priority_auths] couldn't connect to redis pub: ", redis_sub_connect: "...", ... }`

What do you think?

File: xc/cache.lua (hunk 1 above)
Comment (mayorova) [reply xc/cache.lua]: @davidor Can you help with the test please?

File: xc/priority_auths.lua (hunk 2 above)
Comment (mayorova) [reply xc/priority_auths.lua:14]: To me it looks like overcomplicating the things a bit.
Also, for troubleshooting, I actually find it useful to just take the error message found in the logs, and search for where it appears in the code - makes it easy to trace the issue.

File: xc/cache.lua (hunk 1 above)
Comment (davidor) [reply xc/cache.lua]: Sure, I'll take care of that

File: xc/priority_auths.lua (hunk 2 above)
Comment (davidor) [reply xc/priority_auths.lua:14]: It's OK for now, but later we might want to refactor this and extract parts of the logging functionality to its own module.

File: xc/redis_pool.lua [hunk 6]
```diff
@@ -31,7 +32,7 @@ function _M.acquire()
 
   conn:set_timeout(_M.timeout)
 
-  local ok, err = conn:connect(_M.host, _M.port)
+  local ok, err = conn:connect(threescale_utils.resolve(_M.host, _M.port))
```
Comment (unleashed) [inline xc/redis_pool.lua]: can't this fail? if it does, it just blows up the request?

**Comment by davidor** [comment]
@daria : I squashed your third commit as it was only fixing things introduced in the second commit. I also deleted `spec_helper.lua` and `.busted` as they are not needed because of the way I tested the logging part in the commit I added.

File: xc/redis_pool.lua (hunk 6 above)
Comment (unleashed) [reply xc/redis_pool.lua]: I mean the resolving in particular.

Comment (davidor) [reply xc/redis_pool.lua]: Good point. We should check this.

Comment (davidor) [reply xc/redis_pool.lua]: I implemented some error handling and tested it. Check the 2 new commits that I added.

**Comment by unleashed** [comment]
The `ngx.log` directives are basically actively fought against. Logging is something that calling code should handle. That's why we have a mechanism for notifying errors to callers. In fact, some log calls are placed right after specifying in comments that errors are returned in a return parameter.

Now admittedly the mechanism is probably not being used in all places that it should be, and some errors are recoverable and this is not taken into account (or because they are recoverable, information that would be good to have is lost, ie. if cache fails to contact Redis but somehow later priority auth works).

I think using the error reporting mechanism we have here will not solve all needs, especially if we want warnings and not hard errors, so either we put this on hold or we commit to reviewing the mechanism and clean these calls later on.

Such a mechanism could be a function for log information (including log level and potentially callsite info, no matter whether actual logging is performed at that time or later) and/or adding more information to the errors entry in the returned data to account for that extra information from non-fatal circumstances.

Either way the rest of the PR looks good.

**Review by unleashed — CHANGES_REQUESTED** [review]
Sorry, blocking until the performance regression caused by resolving is fixed. Resolving should not be a perf regression if proper TTL handling is done.

Edit: @davidor to make sure there is no such regression.
Edit2: no performance regression! 👍

**Comment by davidor** [comment]
I agree @unleashed . I think we should review the logging mechanism. However, the log calls that @mayorova added can be useful when debugging. So I think we can merge this to give something to our users until we implement a better logging mechanism with the characteristics that you described in your comment.

Regarding the performance regression, I'll run some tests tomorrow to be sure.

**Review by davidor — APPROVED** [review]
It works well :+1:

**Review by unleashed — APPROVED** [review]

**Pull request merged**